---
title: "Prediction HiPSC"
author: "Jose Alquicira Hernandez"
date: "26 Jun 2017"
output: html_document
---

# Load libraries

```{r load_libraries, results='hide', message=FALSE, warning=FALSE}
# module load R/3.4.0
suppressMessages(library('scPrediction'))
library("tidyverse")
library("cowplot")
library("RColorBrewer")
```


```{r command_line_args}
argv <- commandArgs(TRUE)
f.path <- argv[1]
positive.class <- argv[2]
```


# Read and process input data

```{r read_data}
# f.path <- "/Users/j.alquicira/Documents/powell_lab/projects/SingleCell_Prediction/results/2017-05-22_IPSCAnalysis/test/"

hipsc <- readRDS(paste0(f.path, 'Expression_data_HiPSC_5day0Samples.RDS'))

expression.data <- as.matrix(hipsc[[1]])
cell.sample.data <- hipsc[[2]]
groups <- cell.sample.data$Cluster

# expression.data <- readRDS(paste0(f.path,'ori_dat_sample.RDS'))
# groups <- readRDS(paste0(f.path,'my.clusters.RDS'))
```


- `expression.data` corresponds to a gene expression matrix. Rows are *genes* and columns are *cells*
- `cell.clusters` is a vector. Each value corresponds to one cluster. The length of the vector is the total of cell


```{r format_gene_names}
rownames(expression.data) <- gsub("_.*", '', rownames(expression.data))
```

```{r set_negative_class}
negative.class <- paste0(sort(unique(groups)[!unique(groups) %in% positive.class]), collapse = "")
```


```{r read_DEdata}
# Read differentially expressed data
DE.genes <- read.table(paste0(f.path,'result_table_DESeq_', positive.class, '_vs_cluster', negative.class, '_significant10x15K.txt'), header = TRUE)

# Gets gene names from `DE_result` and gets indexes of differential expressed genes in `expression.data` 
DE.gene.names <- gsub("_.*", '',  DE.genes$id)
DE.gene.idx <- which(rownames(expression.data) %in% DE.gene.names)
```



# Starts prediction

```{r set_positive_class}
groups <- ifelse(groups == positive.class, groups, negative.class)
```



```{r set_params_TrainPredict}
seeds.select <- c(18963,11835,38268,4071,515,38489,43612,37577,24484,20562,28354,30815,29137,29246,23762,15286,33605,26739,22919,32596,14373,46398,46064,39920,24110,20294,41148,48654,34029,25359,3016,22720,43078,45935,7979,430,3033,27125,10152,38860,13293,4566,31483,12965,21939,31440,31397,2100,19793,15853,43780,40804,9741,44354,13126,40070,40929,21237,38687,8988,19795,44104,39181,49170,48984,39446,22446,37099,11091,47020,45130,16437,42807,42560,44595,33712,34490,38548,30927,47522,41510,4189,5421,36331,39808,16249,40572,41146,41607,11705,12381,4728,31350,25172,4763,4181,17638,46408,39838,16460)

seeds.compare <- c(9351,26634,16496,4146,30796,47316,37798,2283,37324,12043,20234,28458,13545,5507,39962,2151,24861,36024,1727,4813,44701,32887,32101,38658,28058,4139,36523,24541,7204,46413,21999,9407,10986,13552,35608,1484,40106,18300,44515,32833,48557,19541,22954,32915,33148,9983,5081,28216,49692,20618,21360,30864,16166,35157,10310,25301,36652,44388,30874,10472,24462,36189,45468,19067,35199,29697,18077,7920,27799,37463,29394,6106,22739,10333,2933,30697,764,34893,27751,12255,28325,42947,31440,39348,48143,27205,341,31589,45188,111,759,5723,36586,36847,33712,47969,3351,23386,35620,40907)


dataset.params <- list(groups = groups,
                       positive.class = as.character(positive.class),
                       features = DE.gene.idx,
                       expression.data = expression.data,
                       seed.select = seeds.select,
                       seed.compare = seeds.compare)

training.params <- list(method = 'cv',
                        number = 5,
                        seed = 92,
                        returnResamp = 'none',
                        returnData = FALSE)

methods <- c('rf')
```


```{r make_predictions}
res <- TrainPredict(dataset.params, training.params, methods = methods, n.rep = 10, n.cores = 10)
```


```{r get_accuracy}
res %>%
  lapply(function(b.replicate){ # Iterates bootstrap replicate
    mapply(function(model, model.name){ # Iterates model
      data.frame(
          Model = model.name,
          Accuracy = model$conf.mat$overall["Accuracy"],
          Sensitivity = model$conf.mat$byClass["Sensitivity"],
          Specificity = model$conf.mat$byClass["Specificity"],
          Kappa = model$conf.mat$overall["Kappa"]
        )
    }, b.replicate, names(b.replicate), SIMPLIFY = FALSE) %>% Reduce(rbind, .) # Iterates model
  }) %>% Reduce(rbind, .) -> accuracy # Iterates bootstrap replicate
```

```{r save_results}
cat("Saving metrics...\n")
saveRDS(accuracy, file = paste0("methods_accuracy_cluster", positive.class, ".RDS"))
cat("Results saved...\n")
```


```{r sets_model_plotting_order}
model.order <- c('glm',
                 'bayesglm', 
                 'ridge', 'elasticNet', 'lasso',
                 'earth', 
                 'svmLinear', 'svmPoly', 
                 'rpart', 
                 'rf', 
                 'nb', 
                 'nnet',
                 'dnn')
```

```{r gather_accuracy_data}
metrics <- gather(data = accuracy, key = "Metric", value = "Value", 2:4)

metrics %>% 
  filter(Metric != "Kappa") -> metrics
```

```{r plot_metrics}
p <- ggplot(metrics, aes(x = Metric, y = Value, fill = Metric)) +
  geom_boxplot(position = "dodge") +
  ggtitle("Performance metrics: accuracy, sensitivity and specificity", subtitle = paste("Cluster", positive.class)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_brewer(palette = "Set1")
```


```{r save_metrics_plot}
ggsave(filename = paste0("metrics_cluster_", positive.class,".png"),
       plot = p,
       width = 8,
       height = 7,
       device = "png",
       dpi = 350)
```


```{r display_session_info}
options(width = 120)
devtools::session_info()
```